### ISSUE TITLE
Cache management for custom URI scheme

### ISSUE THREAD

**Issue by mikepantazoglou** [Metrological/WebKitForWayland#1091]
While testing our HbbTV implementation in WebKit, we noticed that the browser is caching the files retrieved by our custom URI (dvb://) scheme request handler. The content of those files is then changed in the DSM-CC carousel, yet the next time the application requests them there is no call to our request handler and the old file content is returned.

Is there any way to disable caching for a custom URI scheme? Or else, what would be the best way to deal with such cases?

**Comment by magomez** [comment]
@mikepantazoglou, is this 2.28?

**Comment by mikepantazoglou** [comment]
@magomez Yes, specifically it is wpe-2.28.1.

**Comment by magomez** [comment]
> While testing our HbbTV implementation in WebKit, we noticed that the browser is caching the files retrieved by our custom URI (dvb://) scheme request handler. The content of those files is then changed in the DSM-CC carousel, yet the next time the application requests them there is no call to our request handler and the old file content is returned.

Are you saying that the whole dvb:// request result is cached? or that the files retrieved by your custom handler are cached?  I'm trying to reproduce the problem but for me the custom URI handler is always executed, which means that the dvb:// request result is not cached, but regenerated for each request. Is that what happens to you? Or is it really that the custom handler is not executed at all?

I'm attaching a small logging patch here. It will just print how each URI will be loaded. It should show whether your requests are going through the custom URI handler or through the cache. Could you reproduce the problem with the patch and send the result back to me?

[1091-log.patch.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/11543799/1091-log.patch.txt)

**Comment by mikepantazoglou** [comment]
It seems that the files retrieved by the custom handler are cached. If the same dvb:// uri is given to the browser for a second time, the custom handler is not called. Instead, the previously retrieved file is returned. The problem in our case is that, the file has in the mean time changed in the DSM-CC carousel, and the new content should be fetched instead of the original once, which seems to have been cached by the browser.

I will apply the patch and give you further feedback as requested.

**Comment by mikepantazoglou** [comment]
Please find attached the logs corresponding to the test with your patch applied. Our findings are as follows:

- The extra logs from the WebLoaderStrategy are only printed out for the HTTP request that is made to load the test page, before any subsequent call to the DSM-CC is made (see lines 408-409). This means our custom URI handler does not end up calling the WebLoaderStrategy code.
- The test involves two calls to the DSM-CC, which are carried out by our custom URI handler. However, the logs show that the URI handler is called only the first time (see [ORBPlatform_Dsmcc.cpp]::[Dsmcc_RequestFile] in line 509).

I don't know if it helps, but perhaps it's worth mentioning that, in our previous integration with wpe-2.22, we ended up patching WebKit to disable patching for URLs with dvb protocol. Here is an excerpt from the patch:

```
--- a/Source/WebCore/xml/XMLHttpRequest.cpp
+++ b/Source/WebCore/xml/XMLHttpRequest.cpp
@@ -605,7 +605,7 @@ ExceptionOr<void> XMLHttpRequest::createRequest()
     options.filteringPolicy = ResponseFilteringPolicy::Enable;
     options.sniffContentEncoding = ContentEncodingSniffingPolicy::DoNotSniff;
 
-    if (responseType() == ResponseType::Arraybuffer || getenv("WPE_DISABLE_XHR_RESPONSE_CACHING")) {
+    if (m_url.protocolIsDVB() || responseType() == ResponseType::Arraybuffer || getenv("WPE_DISABLE_XHR_RESPONSE_CACHING")) {
         options.dataBufferingPolicy = DataBufferingPolicy::DoNotBufferData;
         options.cachingPolicy = CachingPolicy::DisallowCaching;
         request.setCachePolicy(ResourceRequestCachePolicy::DoNotUseAnyCache);
```

The `protocolIsDVB()` method used here was also part of the complete patch along with other stuff.

[logs.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/11545553/logs.txt)

**Comment by magomez** [comment]
Oh, you're not loading a page with a custom URI handler. You're loading a normal HTTP/S page that uses the defined custom URI handler to load XMlHttpRequests. It's different than I thought.
So what happens here is not that the page that you're loading is dvb:// and it doesn't trigger the custom URI handler. What happens is that the page gets loaded, and those XMLHttpRequests are the ones that are being cached, and that's why the custom URI handler is not triggered.

Then yes, the fix that you're looking for is to avoid caching XMLHttpRequest responses (like that change you did for 2.22).

**Comment by mikepantazoglou** [comment]
In general there are two use cases for DSM-CC:

1. Load a dvb:// URL directly in the browser
2. Load an http(s):// URL in the browser, and the page then issues various XHR's with dvb:// URLs

Both cases end up calling our custom URI handler. In both cases, we would like the cache to be disabled for any dvb:// URL requests.

About the fix in wpe-2.22, are you suggesting that we patch wpe-2.28 in the same way to address the aforementioned use case no.2?

**Comment by magomez** [comment]
> In general there are two use cases for DSM-CC:
> 
> 1. Load a dvb:// URL directly in the browser
> 2. Load an http(s):// URL in the browser, and the page then issues various XHR's with dvb:// URLs
> 
> Both cases end up calling our custom URI handler. In both cases, we would like the cache to be disabled for any dvb:// URL requests.

The first case is the one that I've testing, and for me browser always calls the custom handler, so it should not cache the results.

The second case is the problematic one, and it happens because XMLHttpRequest is caching the results of the queries, and that's why the custom handler is only called once.
 
> About the fix in wpe-2.22, are you suggesting that we patch wpe-2.28 in the same way to address the aforementioned use case no.2?

Yes. We can add the same check to disable caching for the dvb handler, or even all the custom handlers (as others may show the same problem). I'll craft a patch for this.

**Comment by magomez** [comment]
I've created a patch to fix this that tries to be a bit generic. For this, I've added a new env variable WPE_DISABLE_XHR_RESPONSE_CACHING_FOR_PROTOCOLS which should contain a comma separated list of protocols for which the XMLHttpRequest cache should be disabled. In your case, something like this should work:

WPE_DISABLE_XHR_RESPONSE_CACHING_FOR_PROTOCOLS=dvb

I'm attaching a patch that checks the value of that env var and disables caching for those protocols. @mikepantazoglou could you apply the patch and export the env var to check whether it fixes the problem for you? If it works properly, I'll commit it.

[1091.patch.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/11600176/1091.patch.txt)

**Comment by mikepantazoglou** [comment]
@magomez Thank you for the patch. We applied it and successfully repeated our tests. I confirm it works properly at least in our cases.

**Comment by magomez** [comment]
@mikepantazoglou I've just pushed 64c9073dd91e530e12a3da5056ec5f9c8c1e3983 with the fix to the wpe-2.28 branch. Please close this once the fix is verified. Thanks!

**Comment by mikepantazoglou** [comment]
Having successfully tested the fix, from our perspective this issue may now be closed.

**Comment by magomez** [comment]
@emutavchi I've received a notification about this comment from you about the fix for this in commit 64c9073dd91e530e12a3da5056ec5f9c8c1e3983. I can't find the comment in github using the links in the email, so I don't know if it's a github error or if maybe you removed the comment.

_This breaks the exiting configuration in webkit browser plugin https://github.com/rdkcentral/rdkservices/blob/main/WebKitBrowser/WebKitImplementation.cpp#L2194_

In theory the commit should not affect the previous behavior related to the WPE_DISABLE_XHR_RESPONSE_CACHING env var. The value of that var is checked in order to disable all caching. Then, if that env var is not defined, it's when the WPE_DISABLE_XHR_RESPONSE_CACHING_FOR_PROTOCOLS var is used to disable caching by protocol.

At least that's what should happen, but maybe I made a mistake. Can you open a new issue if the feature has indeed been broken, please?
